benches/poseidon_benchmarks.rs

```diff
@@ -7,48 +7,103 @@ use criterion::black_box;
 
-/// This function allows us to bench the whole process of
-/// digesting a vec of `Scalar`. Since the hasher instantiation
-/// time matters here, and is optimized or even non-existing on other
-/// implementations, that makes us bench the whole process instead of
-/// just benching the `hasher.result()` fn.
-#[inline]
-fn digest_one() -> () {
-    let mut hasher = Hash::new();
-    let s = Scalar::random(&mut thread_rng());
-    hasher.input(s).unwrap();
-    hasher.result();
-}
+mod scalar_benches {
+    use super::*;
+    
+    /// This function allows us to bench the whole process of
+    /// digesting a vec of `Scalar`. Since the hasher instantiation
+    /// time matters here, and is optimized or even non-existing on other
+    /// implementations, that makes us bench the whole process instead of
+    /// just benching the `hasher.result()` fn.
+    #[inline]
+    fn digest_one() -> () {
+        let mut hasher = Hash::new();
+        let s = Scalar::random(&mut thread_rng());
+        hasher.input(s).unwrap();
+        hasher.result();
+    }
 
-#[inline]
-fn digest(n: usize) -> () {
-    let mut hasher = Hash::new();
-    let s: Vec<Scalar> = (0..n).map(|_| Scalar::random(&mut thread_rng())).collect();
-    hasher.inputs(s).unwrap();
-    hasher.result();
-}
+    #[inline]
+    fn digest(n: usize) -> () {
+        let mut hasher = Hash::new();
+        let s: Vec<Scalar> = (0..n).map(|_| Scalar::random(&mut thread_rng())).collect();
+        hasher.inputs(s).unwrap();
+        hasher.result();
+    }
 
-pub fn hash_one_scalar(c: &mut Criterion) {
-    c.bench_function("Hashing single `Scalar`", |b| b.iter(|| digest_one()));
-}
+    pub fn hash_one_scalar(c: &mut Criterion) {
+        c.bench_function("Hashing single `Scalar`", |b| b.iter(|| digest_one()));
+    }
+
+    pub fn hash_two_scalars(c: &mut Criterion) {
+        c.bench_function("Hashing two `Scalar`", |b| b.iter(|| digest(black_box(2))));
+    }
 
-pub fn hash_two_scalars(c: &mut Criterion) {
-    c.bench_function("Hashing two `Scalar`", |b| b.iter(|| digest(black_box(2))));
+    pub fn hash_four_scalars(c: &mut Criterion) {
+        c.bench_function("Hashing four `Scalar`", |b| b.iter(|| digest(black_box(4))));
+    }
 }
 
-pub fn hash_four_scalars(c: &mut Criterion) {
-    c.bench_function("Hashing four `Scalar`", |b| b.iter(|| digest(black_box(4))));
+mod lc_benches {
+    use super::*;
+
+    /// This function allows us to bench the whole process of
+    /// digesting a vec of `LinearCombination`. We put together the process
+    /// of creating the linearcombination array + digesting the info to be able
+    /// to compare the results against the unoptimized version.
+    #[inline]
+    fn digest_one() -> () {
+        let mut hasher = Hash::new();
+        let s = LinearCombination::from(Scalar::random(&mut thread_rng()));
+
+        let mut verifier_transcript = Transcript::new(b"");
+        let mut verifier = Verifier::new(&mut verifier_transcript);
```

Comment:
As discussed, I don't think the creation of the `transcript` and `verifier` is the culprit: it's just the constraint system that is slower.

One thing you could do, if you want, is removing the creation of the CS from the `digest`* function, so that it won't impact the final result. However, it would negligible.

Something like:
```rust
    fn digest_one(cs: &mut dyn ConstraintSystem) -> () {
        let mut hasher = Hash::new();
        let s = LinearCombination::from(Scalar::random(&mut thread_rng()));

        hasher.input_lc(s).unwrap();
        hasher.result_gadget(cs).unwrap();
    }


    pub fn hash_one_lc(c: &mut Criterion) {
        let mut transcript = Transcript::new(b"");
        let mut cs = Verifier::new(&mut transcript);

        c.bench_function("Hashing single `LinearCombination`", 
            |b| b.iter(|| digest_one(&mut cs)));
    }
```
So it will impact only one time for each iteration group, instead of for each iteration.

Reply:
Did that and also checked the benchmarks for creating a `Verifier`.

As discussed, the time taken is aprox: `625.21 ns` so it's not affecting anything.